rmf_fleet_adapter/src/rmf_fleet_adapter/agv/FleetUpdateHandle.cpp

```diff
@@ -974,6 +978,76 @@ void FleetUpdateHandle::Implementation::add_standard_tasks()
     node->clock());
 }
 
+//==============================================================================
+FleetUpdateHandle& FleetUpdateHandle::add_performable_action(
+  const std::string& category)
+{
+  if (category.empty())
+    return *this;
+
+  // TODO(YV): Consider storing the validator and deserializer within
+  // TaskDeserialization
+  auto validator = _pimpl->deserialization.make_validator_shared(
+    schemas::event_description_PerformAction);
+
+  const auto deserializer =
+    [validator, place = _pimpl->deserialization.place](
+      const nlohmann::json& msg) -> DeserializedEvent
+    {
+      validator->validate(msg);
+      try
+      {
+        const nlohmann::json action =
+          nlohmann::json::parse(msg["action"].get<std::string>());
```

Comment:
I don't think this is a good way to use the `nlohmann` library or JSON more generally. What this is literally saying is that the `action` field will refer to a string, and that string will contain an entire JSON message inside of it. So that might look something like this:

```
"action": "{
  \"clean\": {
    \"zone\": \"terminal_3\"
  }
}"
```

It would become a nightmare of escaping quotes and a very confusing structure.

I recommend the following schema instead:

```
"perform_action": {
  "category": "some_simple_string",
  "description": {
    "any": 0
    "details": { },
    "about": [ ],
    "action: " "
  },
  "expected_finish_location": <optional>,
  "unix_millis_action_duration_estimate": <optional>,
  ... any other optional task planner details ...
}
```

The `add_performable_action` function will have this signature instead:

```
FleetUpdateHandle& FleetUpdateHandle::add_performable_action(
  const std::string& category,
  ConsiderRequest consider);
```

We will maintain a map from `category` keys to `consider` callbacks. When a `perform_action` event comes in, we will do a look up on that map. If the lookup fails, we reject the task. If the look up succeeds, we trigger the `consider` callback and then accept or reject based on its response.

The `ActionExecutor` signature would look like this instead:

```
 using ActionExecutor =
    std::function<void(
      const std::string& category,
      const nlohmann::json& description,
      ActionCompleted completed)
```

This should make it very easy for users to put in their own arbitrary JSON details for the action while still allowing the fleet adapter to do the heavy lifting in weeding out irrelevant task requests.

Reply:
Thanks for the feedback. I can see how this is much easier for the user. I have a couple of questions on the schema
1) What is the purpose of the `any`, `detail` and `about` fields?
2) I assume the intention is for the entire `description` JSON block to be passed into `PerformAction::Description::make()` as the `action`. Is this correct?

Reply:
The `any`, `details`, `about`, `this`, `action` are just examples of how a user can fill the `description` JSON with any kind of data they want. You can totally disregard them, they're not actually part of the schema, just placeholders.

> I assume the intention is for the entire description JSON block to be passed into PerformAction::Description::make() as the action. Is this correct?

That's right, the `PerformAction::Description` would need to have some kind of `action_message` field (or something similarly named) to store the message that gets passed to the robot command handle.